### PR TITLE
[Filesystem] Add test to prevent regression when using array|resource with dumpFile

### PR DESCRIPTION
**Pull request by thePanz**
| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT

### CODE REVIEW

**Comment by ro0NL** [comment]
See https://github.com/symfony/symfony/pull/20980#issuecomment-269467562 :)

**Comment by thePanz** [comment]
Damn, didn't find that PR before opening this one @ro0NL! 
So I guess the only way to use a `resource` with `Filesystem::dumpFile()` is to copy its code somewhere else :(

**Comment by nicolas-grekas** [comment]
Should we reconsider since this works? If we really want a string, we should maybe throw a deprecation? That will force us to consider which alternative we should provide to ppl hit by the notice (and maybe make use realize it's simpler to just accept what already works?)

**Comment by thePanz** [comment]
@nicolas-grekas Rebased, cleaned-up and pushed

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 1]
```diff
@@ -1201,6 +1201,31 @@ public function testDumpFile()
         $this->assertStringEqualsFile($filename, 'bar');
     }
 
+    public function testDumpFileWithArray()
+    {
+        $filename = $this->workspace.DIRECTORY_SEPARATOR.'foo'.DIRECTORY_SEPARATOR.'baz.txt';
```
[the remark this answers is not in the document]
Comment (thePanz) [reply src/Symfony/Component/Filesystem/Tests/FilesystemTest.php]: Changed!

**Review by nicolas-grekas — APPROVED** [review]
yes, I changed my mind: there are two ways here: either "officialize" what already works (this PR), or deprecate what shouldn't work. The least effort is this PR, and the downside (making an implementation that doesn't rely on file_put_contents more complex) hypothetical.

**Comment by ro0NL** [comment]
Note in 3.4 there's also `appendToFile()` which should be updated as well IMHO

**Comment by fabpot** [comment]
I'm 👎 on this change for the reasons expressed in the linked PR.

**Comment by nicolas-grekas** [comment]
I reverted the docblock change but kept the test case so that we don't accidentally break BC.
Removing support for array|resource would need a deprecation layer.

:+1:

**Review by xabbuh — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @thePanz.

**Pull request merged**